core/src/apps/solana/transaction/instruction.py

```diff
@@ -0,0 +1,141 @@
+    ):
+        parsed_account = {}
+        for i, account_template in enumerate(accounts_template):
+            if i >= len(accounts):
+                if account_template.optional:
+                    continue
+                else:
+                    raise ValueError  # "Account is missing
+
+            parsed_account[account_template.name] = accounts[i]
+        return parsed_account
+
+    def __init__(
+        self,
+        instruction_data: bytes,
+        program_id: str,
+        accounts: list[Account],
+        instruction_id: int,
+        property_templates: list[PropertyTemplate],
+        accounts_template: list[AccountTemplate],
+        ui_properties: list[UIProperty],
+        ui_name: str,
+        is_program_supported: bool = True,
+        is_instruction_supported: bool = True,
+        supports_multisig: bool = False,
+        is_deprecated_warning: str | None = None,
+    ) -> None:
+        self.program_id = program_id
+        self.instruction_id = instruction_id
+
+        self.property_templates = property_templates
+        self.accounts_template = accounts_template
+
+        self.ui_name = ui_name
+
+        self.ui_properties = ui_properties
+
+        self.is_program_supported = is_program_supported
+        self.is_instruction_supported = is_instruction_supported
+
+        self.is_deprecated_warning = is_deprecated_warning
+
+        self.instruction_data = instruction_data
+        self.accounts = accounts
+
+        if self.is_instruction_supported:
+            self.parsed_data = self.parse_instruction_data(
+                instruction_data, property_templates
+            )
+
+            self.parsed_accounts = self.parse_instruction_accounts(
+                accounts, accounts_template
+            )
+
+            self.multisig_signers = accounts[len(accounts_template) :]
+
+            if self.multisig_signers and not supports_multisig:
+                raise ValueError  # Multisig not supported
+
+    def __getattr__(self, attr: str) -> Any:
+        if attr in self.parsed_data:
+            return self.parsed_data[attr]
+        if attr in self.parsed_accounts:
+            return self.parsed_accounts[attr]
+
+        raise AttributeError  # Attribute not found
+
+    def get_property_template(self, property: str) -> PropertyTemplate:
+        for property_template in self.property_templates:
+            if property_template.name == property:
+                return property_template
+
+        raise ValueError  # Property not found
+
+    def get_account_template(self, account_name: str) -> AccountTemplate:
+        for account_template in self.accounts_template:
+            if account_template.name == account_name:
+                return account_template
+
+        raise ValueError  # Account not found
+
+    @classmethod
+    def is_type_of(cls, ins: Any) -> TypeGuard[Self]:
+        # gets overridden in `instructions.py` `FakeClass`
+        return False
```

Comment:
`raise NotImplementedError` would be a slightly better choice

Reply:
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-8494caa0738cc8ad43c528cd82094529785133f6a2d7c4747c6ab7a91b6a034eR144